File: app/src/main/java/org/eyeseetea/malariacare/presentation/presenters/OrgUnitProgramFilterPresenter.java
package org.eyeseetea.malariacare.presentation.presenters;

import org.eyeseetea.malariacare.R;
import org.eyeseetea.malariacare.data.database.model.OrgUnitDB;
import org.eyeseetea.malariacare.data.database.model.ProgramDB;
import org.eyeseetea.malariacare.data.database.utils.PreferencesState;
import java.util.List;

public class OrgUnitProgramFilterPresenter {
    View view;

    private List<ProgramDB> programs;
    private List<OrgUnitDB> orgUnits;
    private ProgramDB mProgramDefaultOption;
    private OrgUnitDB mOrgUnitDefaultOption;
    private ProgramDB mSelectedProgramFilter;
    private OrgUnitDB mSelectedOrgUnitFilter;
    private boolean mExclusiveFilter;

    public void attachView(View view){
        this.view = view;
        loadOrgUnits();
        loadPrograms();
    }

    private void loadOrgUnits(){
        orgUnits = OrgUnitDB.list();

        mOrgUnitDefaultOption = new OrgUnitDB(PreferencesState.getInstance().getContext().getResources().getString(
                R.string.filter_all_org_units));

        mSelectedOrgUnitFilter = mOrgUnitDefaultOption;

        orgUnits.add(0, mOrgUnitDefaultOption);

        if (view != null){
            view.renderOrgUnits(orgUnits);
        }
    }

    private void loadPrograms(){
        programs = ProgramDB.list();
        mProgramDefaultOption = new ProgramDB(PreferencesState.getInstance().getContext().getResources().getString(
                R.string.filter_all_org_assessments));

        mSelectedProgramFilter = mProgramDefaultOption;

        programs.add(0, mProgramDefaultOption);

        if (view != null){
            view.renderPrograms(programs);
        }
    }

    public void onProgramSelected(ProgramDB program) {
        if (!mSelectedProgramFilter.equals(program)) {
            mSelectedProgramFilter = program;

            if (mExclusiveFilter)
                unSelectOrgUnit();

            if (view != null) {
                view.notifyProgramFilterChange(mSelectedProgramFilter);
            }
        }
    }

    private void unSelectOrgUnit() {
        mSelectedOrgUnitFilter = mOrgUnitDefaultOption;

        if (view != null) {
            view.unSelectOrgUnitFilter();
        }
    }

    private void unSelectProgram() {
        mSelectedProgramFilter = mProgramDefaultOption;

        if (view != null) {
            view.unSelectProgramFilter();
        }
    }

    private void notifySelectOrgUnit() {
        int indexToSelect = orgUnits.indexOf(mSelectedOrgUnitFilter);

        if (view != null) {
            view.selectOrgUnitFilter(indexToSelect);
        }
    }

    private void notifySelectProgram() {
        int indexToSelect = programs.indexOf(mSelectedProgramFilter);;

        if (view != null) {
            view.selectProgramFilter(indexToSelect);
        }
    }

    public void onOrgUnitSelected(OrgUnitDB orgUnit) {
        if (!mSelectedOrgUnitFilter.equals(orgUnit)) {
            mSelectedOrgUnitFilter = orgUnit;

            if (mExclusiveFilter)
                unSelectProgram();

            if (view != null) {
                view.notifyOrgUnitFilterChange(mSelectedOrgUnitFilter);

            }
        }
    }

    public ProgramDB getSelectedProgramFilter() {
        return mSelectedProgramFilter;
    }

    public OrgUnitDB getSelectedOrgUnitFilter() {
        return mSelectedOrgUnitFilter;
    }

    public void setExclusiveFilter(boolean exclusiveFilter) {
        mExclusiveFilter = exclusiveFilter;
    }

    public void changeSelectedFilters(String programUidFilter, String orgUnitUidFilter) {
        ProgramDB programToSelect = ProgramDB.getProgram(programUidFilter);
        OrgUnitDB orgUnitToSelect = OrgUnitDB.getOrgUnit(orgUnitUidFilter);

        if (programToSelect == null)
            programToSelect = mProgramDefaultOption;

        if (orgUnitToSelect == null )
            orgUnitToSelect = mOrgUnitDefaultOption;


        onOrgUnitSelected(orgUnitToSelect);
        onProgramSelected(programToSelect);

        notifySelectOrgUnit();
        notifySelectProgram();
    }

    public void resetAllFilters() {
        onOrgUnitSelected(mOrgUnitDefaultOption);
        onProgramSelected(mProgramDefaultOption);
    }

    public interface View{
        void renderPrograms(List<ProgramDB> programs);
        void renderOrgUnits(List<OrgUnitDB> orgUnits);
        void notifyProgramFilterChange(ProgramDB programFilter);
        void notifyOrgUnitFilterChange(OrgUnitDB orgUnitFilter);
        void unSelectOrgUnitFilter();
        void unSelectProgramFilter();

        void selectOrgUnitFilter(int indexToSelect);
        void selectProgramFilter(int indexToSelect);
    }
}

File: app/src/main/java/org/eyeseetea/malariacare/presentation/presenters/OrgUnitProgramFilterPresenter.kt
package org.eyeseetea.malariacare.presentation.presenters

import org.eyeseetea.malariacare.domain.entity.OrgUnit
import org.eyeseetea.malariacare.domain.entity.Program
import org.eyeseetea.malariacare.domain.usecase.GetOrgUnitsUseCase
import org.eyeseetea.malariacare.domain.usecase.GetProgramsUseCase
import org.eyeseetea.malariacare.presentation.boundary.Executor

class OrgUnitProgramFilterPresenter(
    private val executor: Executor,
    private val getOrgUnitsUseCase: GetOrgUnitsUseCase,
    private val getProgramsUseCase: GetProgramsUseCase
) {
    internal var view: View? = null

    private var programs: List<Program> = listOf()
    private var orgUnits: List<OrgUnit> = listOf()

    private var programsNames: MutableList<String> = mutableListOf()
    private var orgUnitsNames: MutableList<String> = mutableListOf()

    private var exclusiveFilter: Boolean = false

    private lateinit var allOrgUnitText: String
    private lateinit var allProgramsText: String

    private lateinit var selectedProgramName: String
    private lateinit var selectedOrgUnitName: String

    lateinit var selectedUidProgram: String
    lateinit var selectedUidOrgUnit: String

    fun attachView(view: View, allOrgUnitText: String, allProgramsText: String) {
        this.view = view

        this.allOrgUnitText = allOrgUnitText
        this.allProgramsText = allProgramsText

        loadMetadata()
    }

    fun onProgramSelected(programName: String) {
        if (selectedProgramName != programName) {
            changeSelectedProgram(programName)

            if (exclusiveFilter && programName != allProgramsText) {
                unSelectOrgUnit()
            }

            notifyProgramFilterChange()
        }
    }

    fun onOrgUnitSelected(orgUnitName: String) {
        if (selectedOrgUnitName != orgUnitName) {
            changeSelectedOrgUnit(orgUnitName)

            if (exclusiveFilter && orgUnitName != allOrgUnitText) {
                unSelectProgram()
            }

            notifyOrgUnitFilterChange()
        }
    }

    fun setExclusiveFilter(exclusiveFilter: Boolean) {
        this.exclusiveFilter = exclusiveFilter
    }

    fun changeSelectedFilters(programUidFilter: String, orgUnitUidFilter: String) {
        val programToSelect =
            programs.firstOrNull { program -> program.uid == programUidFilter }

        val orgUnitToSelect =
            orgUnits.firstOrNull { orgUnit -> orgUnit.uid == orgUnitUidFilter }

        val programNameToSelect = programToSelect?.name ?: allProgramsText
        val orgUnitNameToSelect = orgUnitToSelect?.name ?: allOrgUnitText

        onProgramSelected(programNameToSelect)

        if (exclusiveFilter && programNameToSelect != allProgramsText) {
            unSelectOrgUnit()
            notifyOrgUnitFilterChange()
        } else {
            onOrgUnitSelected(orgUnitNameToSelect)
        }

        notifySelectOrgUnit()
        notifySelectProgram()
    }

    private fun loadMetadata() = executor.asyncExecute {
        loadOrgUnits()
        loadPrograms()
    }

    private fun loadOrgUnits() {
        orgUnits = getOrgUnitsUseCase.execute()
        orgUnitsNames = orgUnits.map { orgUnit -> orgUnit.name } as MutableList<String>

        changeSelectedOrgUnit(allOrgUnitText)

        orgUnitsNames.add(0, selectedOrgUnitName)

        showOrgUnits()
    }

    private fun loadPrograms() {
        programs = getProgramsUseCase.execute()
        programsNames = programs.map { program -> program.name } as MutableList<String>

        changeSelectedProgram(allProgramsText)

        programsNames.add(0, selectedProgramName)

        showPrograms()
    }

    private fun changeSelectedProgram(programName: String) {
        selectedProgramName = programName

        if (selectedProgramName == allProgramsText) {
            selectedUidProgram = ""
        } else {
            selectedUidProgram =
                programs.first { program -> program.name == selectedProgramName }.uid
        }
    }

    private fun changeSelectedOrgUnit(orgUnitName: String) {
        selectedOrgUnitName = orgUnitName

        if (selectedOrgUnitName == allOrgUnitText) {
            selectedUidOrgUnit = ""
        } else {
            selectedUidOrgUnit =
                orgUnits.first { orgUnit -> orgUnit.name == selectedOrgUnitName }.uid
        }
    }

    private fun showOrgUnits() = executor.uiExecute {
        view?.showOrgUnits(orgUnitsNames as List<String>)
    }

    private fun showPrograms() = executor.uiExecute {
        view?.showPrograms(programsNames as List<String>)
    }

    private fun notifyProgramFilterChange() = executor.uiExecute {
        view?.notifyProgramFilterChange(selectedUidProgram)
    }

    private fun notifyOrgUnitFilterChange() = executor.uiExecute {
        view?.notifyOrgUnitFilterChange(selectedUidOrgUnit)
    }

    private fun unSelectOrgUnit() = executor.uiExecute {
        changeSelectedOrgUnit(allOrgUnitText)

        view?.unSelectOrgUnitFilter()
    }

    private fun unSelectProgram() = executor.uiExecute {
        changeSelectedProgram(allProgramsText)

        view?.unSelectProgramFilter()
    }

    private fun notifySelectOrgUnit() = executor.uiExecute {
        val indexToSelect = orgUnitsNames.indexOf(selectedOrgUnitName)

        view?.selectOrgUnitFilter(indexToSelect)
    }

    private fun notifySelectProgram() = executor.uiExecute {
        val indexToSelect = programsNames.indexOf(selectedProgramName)

        view?.selectProgramFilter(indexToSelect)
    }

    interface View {
        fun showPrograms(programs: List<@JvmSuppressWildcards String>)
        fun showOrgUnits(orgUnits: List<@JvmSuppressWildcards String>)
        fun notifyProgramFilterChange(programFilter: String)
        fun notifyOrgUnitFilterChange(orgUnitFilter: String)
        fun unSelectOrgUnitFilter()
        fun unSelectProgramFilter()

        fun selectOrgUnitFilter(indexToSelect: Int)
        fun selectProgramFilter(indexToSelect: Int)
    }
}
